controllers/cluster_controller.go

```diff
@@ -174,6 +174,9 @@ func (r *ClusterReconciler) reconcile(ctx context.Context, cluster *clusterv1.Cl
 
 		errs = append(errs, err)
 	}
+	if !res.Requeue {
+		cluster.Status.ObservedGeneration = cluster.Generation
```

Comment:
IMO ObservedGeneration should be always set.
if possible, for all the controllers, we should have this call in the defer statement that calls patch.